File: C#/Array/RotateImage.cs
/*
48. Rotate Image
Link: https://leetcode.com/problems/rotate-image/

You are given an n x n 2D matrix representing an image, rotate
the image by 90 degrees (clockwise).

You have to rotate the image in-place, which means you have to modify
the input 2D matrix directly. DO NOT allocate another 2D matrix and do the rotation.
*/

public class Solution {
    public void Rotate(int[][] matrix) {
        int n = matrix.Length;
        for (int row=0; row<n/2; row++){
            for (int col=row; col<n-row-1; col++){
                int tmp = matrix[row][col];
                matrix[row][col] = matrix[n-col-1][row];
                matrix[n-col-1][row] = matrix[n-row-1][n-col-1];
                matrix[n-row-1][n-col-1] = matrix[col][n-row-1];
                matrix[col][n-row-1] = tmp;
            }
        }
    }
}

File: Python/Array/RotateImage.py
"""
48. Rotate Image
Link: https://leetcode.com/problems/rotate-image/

You are given an n x n 2D matrix representing an image, rotate
the image by 90 degrees (clockwise).

You have to rotate the image in-place, which means you have to modify
the input 2D matrix directly. DO NOT allocate another 2D matrix and do the rotation.
"""

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        
        # flip diagonally
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
                
        # flip vertically
        for i in range(n):
            for j in range(n//2):
                matrix[i][j], matrix[i][n-j-1] = matrix[i][n-j-1], matrix[i][j]